src/Periphery/Permit2Proxy.sol

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { ISignatureTransfer } from "lifi/Interfaces/ISignatureTransfer.sol";
+import { TransferrableOwnership } from "lifi/Helpers/TransferrableOwnership.sol";
+import { LibAsset, IERC20 } from "lifi/Libraries/LibAsset.sol";
+
+interface IPermit2 {
+    function DOMAIN_SEPARATOR() external view returns (bytes32);
+}
+
+/// @title Permit2Proxy
+/// @author LI.FI (https://li.fi)
+/// @notice Proxy contract allowing gasless (Permit2-enabled) calls to our diamond contract
+/// @custom:version 1.0.0
+contract Permit2Proxy is TransferrableOwnership {
+    string private constant _WITNESS_TYPE_STRING =
+        "Witness witness)TokenPermissions(address token,uint256 amount)Witness(address tokenReceiver,address diamondAddress,bytes diamondCalldata)";
```

Comment:
Basically the idea is the following: 

The full string needs to look like this: 
"PermitWitnessTransferFrom(TokenPermissions permitted,address spender,uint256 nonce,uint256 deadline,Witness witness)TokenPermissions(address token,uint256 amount)Witness(address tokenReceiver,address diamondAddress,bytes diamondCalldata)"

Let's split this up in two parts: 
1) "PermitWitnessTransferFrom(TokenPermissions permitted,address spender,uint256 nonce,uint256 deadline,"
2) "Witness witness)TokenPermissions(address token,uint256 amount)Witness(address tokenReceiver,address diamondAddress,bytes diamondCalldata)"

The first part is always the same since it's given by the Permit2 transferFrom function we are calling. Therefore we dont need this part as we can safely assume it. The witness part differs. It could have any name (doesnt need to be "Witness") and can be any struct pretty much. So after we pass the signature of the PermitWitnessTransferFrom function we need to detail all the structs in alphabetical order. So if instead of Witness I would call my struct Bogus then I would have describe it before the TokenPermissions struct. 
The order is basically: 
- first signature of PermitWitnessTransferFrom including "witness" struct
- then all structs that were used in above in alphabetical order

This will allow the Permit2 contract to correctly decode the data and verify it with the signature to make sure that exactly that data was signed by the user.